SEP-2468: Recommend Issuer (iss) Parameter in MCP Auth Responses (#2468)
modelcontextprotocol/modelcontextprotocol: seps/2468-recommend-issuer-claim-for-auth.md

# SEP-2468: Recommend Issuer (iss) Parameter in MCP Auth Responses

- **Status**: In-Review
- **Type**: Standards Track
- **Created**: 2026-03-25
- **Author(s)**: Emily Lauber (@EmLauber)
- **Sponsor**: @pcarleton
- **PR**: https://github.com/modelcontextprotocol/modelcontextprotocol/pull/2468

## Abstract

This SEP proposes requiring the inclusion and validation of an explicit issuer (iss) parameter in Model Context Protocol (MCP) authorization responses to mitigate authorization mix‑up attacks. By binding authorization responses to a specific authorization server identity, MCP clients can reliably detect and reject responses originating from an unexpected issuer, improving protocol robustness in multi‑identity provider (IdP) environments. This SEP follows the specifications defined in [RFC9207](https://datatracker.ietf.org/doc/rfc9207/)

## Motivation

> The Model Context Protocol increasingly operates in environments where multiple authorization servers, identity providers, and intermediaries coexist. In such environments, OAuth 2.0 mix‑up attacks become a realistic threat. Mix-up attacks are when an attacker causes a client to associate an authorization response with the wrong authorization server, potentially leading to token leakage or privilege escalation.

Reviewer: nit: 2.0 / 2.1 / or omit version number entirely?

```suggestion
The Model Context Protocol increasingly operates in environments where multiple authorization servers, identity providers, and intermediaries coexist. In such environments, OAuth mix‑up attacks become a realistic threat. Mix-up attacks are when an attacker causes a client to associate an authorization response with the wrong authorization server, potentially leading to token leakage or privilege escalation.
```

> # SEP-2468: Recommend Issuer (iss) Parameter in MCP Auth Responses

Reviewer 2: nit (non-blocking): the file slug still says `claim` after the parameter rename and becomes the published docs URL. If we want it consistent, rename to `2468-recommend-issuer-parameter-for-auth.md` and re-run `npm run generate:seps`. Existing SEP slugs are not curated, so also fine to leave as-is.

- **Status**: In-Review
- **Type**: Standards Track
- **Created**: 2026-03-25
- **Author(s)**: Emily Lauber (@EmLauber)
- **Sponsor**: @pcarleton
- **PR**: https://github.com/modelcontextprotocol/modelcontextprotocol/pull/2468

## Abstract

> This SEP proposes requiring the inclusion and validation of an explicit issuer (iss) parameter in Model Context Protocol (MCP) authorization responses to mitigate authorization mix‑up attacks. By binding authorization responses to a specific authorization server identity, MCP clients can reliably detect and reject responses originating from an unexpected issuer, improving protocol robustness in multi‑identity provider (IdP) environments. This SEP follows the specifications defined in [RFC9207](https://datatracker.ietf.org/doc/rfc9207/)

Reviewer 2: ```suggestion
This SEP proposes recommending the inclusion and requiring the validation of an explicit issuer (iss) parameter in Model Context Protocol (MCP) authorization responses to mitigate authorization mix‑up attacks. By binding authorization responses to a specific authorization server identity, MCP clients can reliably detect and reject responses originating from an unexpected issuer, improving protocol robustness in multi‑identity provider (IdP) environments. This SEP follows the specifications defined in [RFC9207](https://datatracker.ietf.org/doc/rfc9207/).
```

> This SEP proposes requiring the inclusion and validation of an explicit issuer (iss) parameter in Model Context Protocol (MCP) authorization responses to mitigate authorization mix‑up attacks. By binding authorization responses to a specific authorization server identity, MCP clients can reliably detect and reject responses originating from an unexpected issuer, improving protocol robustness in multi‑identity provider (IdP) environments. This SEP follows the specifications defined in [RFC9207](https://datatracker.ietf.org/doc/rfc9207/)

Reviewer 2: nit: missing terminal period.

```suggestion
This SEP proposes requiring the inclusion and validation of an explicit issuer (iss) parameter in Model Context Protocol (MCP) authorization responses to mitigate authorization mix‑up attacks. By binding authorization responses to a specific authorization server identity, MCP clients can reliably detect and reject responses originating from an unexpected issuer, improving protocol robustness in multi‑identity provider (IdP) environments. This SEP follows the specifications defined in [RFC9207](https://datatracker.ietf.org/doc/rfc9207/).
```

## Motivation

The Model Context Protocol increasingly operates in environments where multiple authorization servers, identity providers, and intermediaries coexist. In such environments, OAuth mix‑up attacks become a realistic threat. Mix-up attacks are when an attacker causes a client to associate an authorization response with the wrong authorization server, potentially leading to token leakage or privilege escalation.

> OAuth specifications describes two mitigations for mix‑up attacks: requiring issuer (_iss_) parameter or using a unique redirect_uri for each issuer a client interacts with. A unique redirect_uri per issuer is not possible with Client ID Metadata Documents (the recommended registration approach) and is operationally expensive with Dynamic Client Registration. As such, the recommendation is for MCP environments to leverage the issuer mitigation.

Reviewer 2: nit: subject-verb agreement.

```suggestion
OAuth specifications describe two mitigations for mix‑up attacks: requiring issuer (_iss_) parameter or using a unique redirect_uri for each issuer a client interacts with. A unique redirect_uri per issuer is not possible with Client ID Metadata Documents (the recommended registration approach) and is operationally expensive with Dynamic Client Registration. As such, the recommendation is for MCP environments to leverage the issuer mitigation.
```

Requiring an explicit iss parameter in MCP authorization responses provides a simple, interoperable, and well‑understood mechanism to bind responses to the correct authorization server and prevent mix‑up attacks by construction. Since not every authorization server sends the issuer parameter though, this SEP proposes a MUST for clients to validate issuer if provided and a SHOULD for authorization servers supporting MCP scenarios. Future SEPs and releases may change the SHOULD to a MUST.

## Specification

### Issuer Parameter Requirement

MCP authorization servers SHOULD include an issuer (_iss_) parameter in authorization responses, including error responses, as defined in [RFC9207](https://datatracker.ietf.org/doc/html/rfc9207#section-2). Authorization servers that do so MUST advertise it by setting `authorization_response_iss_parameter_supported: true` in their authorization server metadata.

> The _iss_ parameter MUST:

Reviewer 2: ```suggestion
The `iss` parameter MUST:
```

> - Exactly match the issuer identifier advertised via discovery or configuration

Reviewer 2: ```suggestion
- Exactly match the issuer identifier advertised via metadata discovery
```

> - Exactly match the issuer identifier advertised via discovery or configuration

Reviewer 2: nit: "or configuration" is RFC 9207 boilerplate for deployments without metadata; MCP mandates AS metadata discovery so there is no other source.

```suggestion
- Exactly match the issuer identifier advertised via metadata discovery
```

> - Its value MUST be a URL that uses the "https" scheme without any query or fragment components.

Reviewer 2: ```suggestion
- Be a URL that uses the `https` scheme without query or fragment components ([RFC 8414 Section 2](https://datatracker.ietf.org/doc/html/rfc8414#section-2))
```

### Client Validation Requirements

MCP clients MUST validate the _iss_ parameter in authorization responses by:

- Determining the expected issuer for the authorization request
- Comparing the received _iss_ value against the expected issuer
- Rejecting the authorization response if the values do not match exactly

If issuer validation fails, the client MUST treat the response as invalid and abort the authorization flow.

> Clients SHOULD continue to apply all existing OAuth security checks in addition to issuer validation.

Reviewer 2: nit: this SHOULD is unscoped ("all existing OAuth security checks" has no defined scope) and does not appear in the spec diff. Suggest dropping it.

```suggestion
If issuer validation fails, the client **MUST** treat the response as invalid and abort the authorization flow.
```

Discarding `iss` when the server does not advertise support (strict RFC 9207 §2.4 SHOULD)

- RFC 9207 §2.4 recommends that clients SHOULD discard responses carrying `iss` from servers that do not set `authorization_response_iss_parameter_supported`, but explicitly leaves the decision to local policy ("specific guidance is out of scope"). This SEP specifies comparison instead. The recorded issuer always comes from a metadata document the client has already validated per RFC 8414 §3.3, so a present `iss` can be checked against an authentic baseline; rejection on mismatch remains unconditional, so the only behavioral difference is accepting a response whose `iss` matches that baseline — which is not a relaxation. In practice, authorization servers often begin emitting `iss` before their metadata is updated, and discarding in that window would reject legitimate flows without security benefit.

## Backward Compatibility

> The `iss` parameter is additive on the wire. Client validation introduces a behavioral change for hosts whose authorization server advertises `authorization_response_iss_parameter_supported: true` but whose callback handling does not yet pass `iss` to the SDK; those flows will be rejected until the host extracts `iss` from the redirect URI alongside `code`. SDKs are expected to widen callback signatures additively (e.g., an optional `iss` argument) so existing call sites continue to compile. Authorization servers that do not advertise support are unaffected.

Reviewer 2: nit: the new RFC 8414 Section 3.3 metadata-validation MUST in `authorization.mdx` applies to every metadata fetch, not just the `iss` callback path. Worth one sentence here so implementers are not surprised.

```suggestion
The `iss` parameter is additive on the wire. Client validation introduces a behavioral change for hosts whose authorization server advertises `authorization_response_iss_parameter_supported: true` but whose callback handling does not yet pass `iss` to the SDK; those flows will be rejected until the host extracts `iss` from the redirect URI alongside `code`. SDKs are expected to widen callback signatures additively (e.g., an optional `iss` argument) so existing call sites continue to compile. Authorization servers that do not advertise support are unaffected. The accompanying RFC 8414 Section 3.3 metadata-validation requirement restates an existing RFC MUST; clients that were not already enforcing it may surface latent issuer misconfigurations on upgrade.
```

## Security Implications

> This proposal is a mitigation against mix-up attacks; the security considerations for the mechanism itself are documented in [RFC9207 Section 4](https://datatracker.ietf.org/doc/html/rfc9207#section-4). In particular, the mitigation depends on clients establishing the expected issuer before redirecting and on the comparison being byte-exact. See also the MCP [security best practices](/docs/tutorials/security/security_best_practices).

Reviewer 2: nit: RFC 3986 Section 6.2.1 defines this as "simple string comparison" and avoids "byte-exact" (misleading for non-ASCII). The spec diff already uses the RFC term.

```suggestion
This proposal is a mitigation against mix-up attacks; the security considerations for the mechanism itself are documented in [RFC9207 Section 4](https://datatracker.ietf.org/doc/html/rfc9207#section-4). In particular, the mitigation depends on clients establishing the expected issuer before redirecting and on the comparison being an exact simple string comparison. See also the MCP [security best practices](/docs/tutorials/security/security_best_practices).
```

## Reference Implementation

- Go SDK: [modelcontextprotocol/go-sdk#859](https://github.com/modelcontextprotocol/go-sdk/pull/859)
- TypeScript SDK: [modelcontextprotocol/typescript-sdk#1957](https://github.com/modelcontextprotocol/typescript-sdk/pull/1957)

> Both implementations key client validation on the `authorization_response_iss_parameter_supported` metadata flag and validate before transmitting the authorization code to the token endpoint.

Reviewer 2: ```suggestion
Both record the expected issuer before redirect and compare any received `iss`, rejecting on absence only when the server advertises support.
```
